### PR TITLE
Smaller toke

### PR DESCRIPTION
**Pull request by arc**
This branch factors lots of code out of `Perl_yylex()`, which started at 4166 lines. It's now split into multiple functions; the largest remaining one is 2159 lines. There's still plenty of work that could be done here, but I hope this makes it easier to understand the lexer without having to understand the whole thing simultaneously.

### CODE REVIEW

File: toke.c [hunk 1]
```diff
@@ -7555,14 +7566,7 @@ Perl_yylex(pTHX)
...
     case '\\':
-	s++;
-	if (PL_lex_inwhat == OP_SUBST && PL_lex_repl == PL_linestr
-	 && isDIGIT(*s))
-	    Perl_ck_warner(aTHX_ packWARN(WARN_SYNTAX),"Can't use \\%c to mean $%c in expression",
-			   *s, *s);
-	if (PL_expect == XOPERATOR)
-	    no_op("Backslash",s);
-	OPERATOR(REFGEN);
+        return yyl_backslash(aTHX_ s + 1);
```
Comment (atoomic) [inline toke.c:7488]: should not we preserve the `s++` some cases are doing at the end a `s--` but I do not view it there

Comment (arc) [reply toke.c:7488]: `s` is a local variable, so given that we're doing a `return` here, I don't think there's much benefit in updating the local variable before passing it to the callee

File: toke.c [hunk 2]
```diff
@@ -5515,6 +5515,64 @@ yyl_hyphen(pTHX_ char *s)
...
 }
 
+static int
+yyl_subproto(pTHX_ char *s, CV *cv)
+{
+    STRLEN protolen = CvPROTOLEN(cv);
+    const char *proto = CvPROTO(cv);
+    bool optional;
+
+    if (!SvPOK(cv))
```
Comment (atoomic) [inline toke.c]: maybe check it and return earlier?

File: toke.c [hunk 3]
```diff
@@ -7424,58 +7485,9 @@ Perl_yylex(pTHX)
...
-		if (hgv && tmp != KEY_x)	/* never ambiguous */
-		    Perl_ck_warner(aTHX_ packWARN(WARN_AMBIGUOUS),
-				   "Ambiguous call resolved as CORE::%s(), "
-				   "qualify as such or use &",
-				   GvENAME(hgv));
-	    }
-	}
+        if (tmp < 0) {
+            tmp = yyl_secondclass_keyword(aTHX_ s, len, tmp, &orig_keyword, &gv, &gvp);
+        }
```
Comment (ilmari) [inline toke.c]: We don't generally use braces around single-statement `if` bodies.

File: toke.c (hunk 2 above)
Comment (arc) [reply toke.c]: Yeah, this commit actually changed the apparent order of the `SvPOK()`, `CvPROTOLEN()`, and `CvPROTO()` operations — good spot, thank you! I'll squash a change in and force-push.

File: toke.c [hunk 4]
```diff
@@ -7309,18 +7310,18 @@ Perl_yylex(pTHX)
         return yyl_bang(aTHX_ s + 1);
 
     case '<':
+        if (s[1] == '<' && (s == PL_linestart || s[-1] == '\n')
+            && memBEGINs(s+2, (STRLEN) (PL_bufend - (s+2)), "<<<<<"))
```
Comment (atoomic) [inline toke.c:7390]: we are  missing the `PL_expect != XOPERATOR` check there?

File: toke.c [hunk 5]
```diff
@@ -6855,6 +6842,21 @@ Perl_yylex(pTHX)
...
+yyl_try(pTHX_ char initial_state, char *s, STRLEN len,
+        I32 orig_keyword, GV *gv, GV **gvp,
+        U8 formbrack, U32 fake_eof, const bool saw_infix_sigil)
+{
+    char *d;
+    bool bof = FALSE;
+
+    switch (initial_state) {
+    case '}': goto rightbracket;
+    }
```
Comment (atoomic) [inline toke.c:6846]: weird switch, but ok

File: toke.c (hunk 4 above)
Comment (arc) [reply toke.c:7390]: The case where `PL_expect` is `XOPERATOR` (not written with an `else`, but the `if (PL_expect != XOPERATOR)` branch ends with an unconditional `TOKEN()` macro invocation, which does a `return`) had the same code — I deleted four hunks that call `vcs_conflict_marker()`, but added only three.

This change makes sense to me: the location of conflict markers in a file is unrelated to the programmer's intent, and therefore to whether the Perl parser is expecting a term or an operator. So we shouldn't restrict conflict-marker detection to situations with specific parser expectations.

**Comment by Leont** [comment]
I can imagine some of this work being very useful in combination with keywords plugins (which suggests making those functions non-static), but I'm not entirely sure which parts. @leonerd?

File: toke.c (hunk 5 above)
Comment (arc) [reply toke.c:6846]: I was thinking there would end up being other situations where a call would need to jump further down, especially once the trickier bits of what's now `yyl_try()` are tackled. But I'm happy to rework this if you think that would be an improvement?

Comment (atoomic) [reply toke.c:6846]: this is fine as it :-) I was just pointing it, right now it s more a boolean context
but I can also see further usage in the future

File: toke.c (hunk 4 above)
Comment (atoomic) [reply toke.c:7390]: thanks for making it clear

File: toke.c [hunk 6]
```diff
@@ -6337,31 +6365,7 @@ Perl_yylex(pTHX)
...
-		}
-		Aop(OP_ADD);
-	    }
-	    else {
-		if (isSPACE(*s) || !isSPACE(*PL_bufptr))
-		    check_uni();
-		OPERATOR('+');
-	    }
-	}
+	return yyl_plus(aTHX_ s);
```
Comment (ilmari) [inline toke.c]: Stray tab for indentation.

File: toke.c [hunk 7]
```diff
@@ -5615,6 +5615,23 @@ yyl_percent(pTHX_ char *s)
     TERM('%');
 }
 
+static int
+yyl_caret(pTHX_ char *s)
+{
+    char *d = s;
+    bool bof = cBOOL(FEATURE_BITWISE_IS_ENABLED);
```
Comment (ilmari) [inline toke.c]: `const`?

File: toke.c [hunk 8]
```diff
@@ -6107,7 +6107,7 @@ yyl_rightcurly(pTHX_ char *s, U8 formbrack)
 }
 
 static int
-yyl_ampersand(pTHX_ char *s, bool bof)
+yyl_ampersand(pTHX_ char *s)
```
Comment (ilmari) [inline toke.c]: Was this a cleanup that was meant to go in the `yyl_ampersand()` commit?

File: toke.c [hunk 9]
```diff
@@ -7033,28 +7063,11 @@ Perl_yylex(pTHX)
...
-	d = s;
-	if ((bof = FEATURE_BITWISE_IS_ENABLED) && *s == '.')
-	    s++;
-	if (!PL_lex_allbrackets && PL_lex_fakeeof >=
-		(*s == '=' ? LEX_FAKEEOF_ASSIGN : LEX_FAKEEOF_BITWISE)) {
-	    s = d - 1;
-	    TOKEN(0);
-	}
-	BOop(bof ? s == d ? OP_NBIT_OR : OP_SBIT_OR : OP_BIT_OR);
+        return yyl_verticalbar(aTHX_ s + 1);
```
Comment (ilmari) [inline toke.c]: This function does `s--`, should it get `s` instead of `s + 1` and do `s++` itself?

File: toke.c [hunk 10]
```diff
@@ -6592,6 +6592,85 @@ yyl_data_handle(pTHX)
     PL_rsfp = NULL;
 }
 
+static void
+yyl_croak_unrecognised(pTHX_ char *s)
```
Comment (ilmari) [inline toke.c:6600]: The commit message only mentions `yyl_require()`, was this meant to be in a separate commit?

Also, should this be marked `__attribute__noreturn__`?

File: toke.c (hunk 3 above)
Comment (arc) [reply toke.c]: Thanks, changed

File: toke.c (hunk 7 above)
Comment (arc) [reply toke.c]: Thanks, changed

File: toke.c (hunk 8 above)
Comment (arc) [reply toke.c]: Indeed it was. Fixed, thank you!

File: toke.c (hunk 9 above)
Comment (arc) [reply toke.c]: Yes, correct

File: toke.c (hunk 10 above)
Comment (arc) [reply toke.c:6600]: Yes and yes. The noreturn dance is apparently a little more complicated than that, but I've found something suitable to copy.

File: toke.c (hunk 6 above)
Comment (arc) [reply toke.c]: Thanks, changed

**Comment by khwilliamson** [comment]
On 10/21/19 2:39 AM, Aaron Crane wrote:
> *@arc* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In toke.c <https://github.com/Perl/perl5/pull/17189#discussion_r336896258>:
> 
>> +        if (tmp < 0) {
> +            tmp = yyl_secondclass_keyword(aTHX_ s, len, tmp, &orig_keyword, &gv, &gvp);
> +        }
> 
> Thanks, changed
> 

The last time this issue was visited, it was decided that using braces 
or not is at the discretion of the implementer.

I will continue to almost always use braces because it has saved me 
numerous times when later adding a statement to the single one already 
there, so that only one of the two is actually controlled by the 'if'. 
There are supposedly nowadays options that warn over this, and I've seen 
them crop up occasionally, but many times no warning was emitted that 
should have been.
